HarmonyPatches/UI/WarnCrouchWalls.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using BetterSongList.Util;
+using HarmonyLib;
+using TMPro;
+using UnityEngine;
+
+namespace BetterSongList.HarmonyPatches.UI {
+	
+	[HarmonyPatch(typeof(LevelParamsPanel), nameof(LevelParamsPanel.obstaclesCount), MethodType.Setter)]
+	static class WarnCrouchWalls {
+
+		static CancellationTokenSource crouchWallTokenSource = null;
+		
+		static readonly Dictionary<BeatmapKey, bool> crouchWallCache = new Dictionary<BeatmapKey, bool>();
+
+		// For some reason, hooking these two methods doesn't work.
+		// Even though they are the only methods that set the obstacle count text,
+		// our text assignment in POSTFIX is still being overwritten. 
+		// [HarmonyTargetMethods]
+		// static IEnumerable<MethodInfo> TargetMethods() {
+		// 	yield return AccessTools.Method(typeof(StandardLevelDetailView), nameof(StandardLevelDetailView.CalculateAndSetContent));
```

Comment:
This happens because you are not patching the correct method. This is an async method, so you have to patch the state machine. You can use `.GetStateMachineTarget()` from MonoMod.Utils as a nice shortcut to that. Though patching a state machine usually implies a transpiler, so your current Postfix probably won't work as-is.

Reply:
I made this really simple patch, but it straight up gives me an IL compile error. I am not even using a transpiler. I am lost...
```cs
	[HarmonyPatch]
	static class WarnCrouchWallsAsyncPatch {
		[HarmonyTargetMethod]
		static MethodInfo TargetMethod() {
			return AccessTools.Method(typeof(StandardLevelDetailView), nameof(StandardLevelDetailView.CalculateAndSetContent))
				.GetStateMachineTarget();
		}

		static void Postfix(object __instance) {
			Plugin.Log.Error("WarnCrouchWallsAsyncPatch.Postfix");
		}
	}
```
The error:
```
[ERROR @ 02:47:48 | IPA] BetterSongList OnEnable: HarmonyLib.HarmonyException: IL Compile Error ---> System.InvalidProgramException: Invalid IL code in (wrapper dynamic-method) StandardLevelDetailView/<CalculateAndSetContent>d__56:DMD<StandardLevelDetailView+<CalculateAndSetContent>d__56::MoveNext> (StandardLevelDetailView/<CalculateAndSetContent>d__56&): IL_0370: call      0x000000e1
[ERROR @ 02:47:48 | IPA]   at (wrapper managed-to-native) System.RuntimeMethodHandle.GetFunctionPointer(intptr)
[ERROR @ 02:47:48 | IPA]   at System.RuntimeMethodHandle.GetFunctionPointer () [0x00000] in <75633565436c42f0a6426b33f0132ade>:0 
[ERROR @ 02:47:48 | IPA]   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeILPlatform.GetFunctionPointer (System.Reflection.MethodBase method, System.RuntimeMethodHandle handle) [0x00000] in <bbf7055c0fdb4018beccdadfe3e89f44>:0 
[ERROR @ 02:47:48 | IPA]   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeILPlatform.GetNativeStart (System.Reflection.MethodBase method) [0x00019] in <bbf7055c0fdb4018beccdadfe3e89f44>:0 
[ERROR @ 02:47:48 | IPA]   at MonoMod.RuntimeDetour.DetourHelper.GetNativeStart (System.Reflection.MethodBase method) [0x00005] in <bbf7055c0fdb4018beccdadfe3e89f44>:0 
[ERROR @ 02:47:48 | IPA]   at MonoMod.RuntimeDetour.Detour._TopApply () [0x00025] in <bbf7055c0fdb4018beccdadfe3e89f44>:0 
[ERROR @ 02:47:48 | IPA]   at MonoMod.RuntimeDetour.Detour._RefreshChain (System.Reflection.MethodBase method) [0x00151] in <bbf7055c0fdb4018beccdadfe3e89f44>:0 
[ERROR @ 02:47:48 | IPA]   at MonoMod.RuntimeDetour.Detour.Apply () [0x00053] in <bbf7055c0fdb4018beccdadfe3e89f44>:0 
[ERROR @ 02:47:48 | IPA]   at MonoMod.RuntimeDetour.Detour..ctor (System.Reflection.MethodBase from, System.Reflection.MethodBase to, MonoMod.RuntimeDetour.DetourConfig& config) [0x002ec] in <bbf7055c0fdb4018beccdadfe3e89f44>:0 
[ERROR @ 02:47:48 | IPA]   at (wrapper dynamic-method) MonoMod.RuntimeDetour.ILHook+Context.DMD<MonoMod.RuntimeDetour.ILHook+Context::Refresh>(MonoMod.RuntimeDetour.ILHook/Context)
[ERROR @ 02:47:48 | IPA]   at (wrapper dynamic-method) MonoMod.Utils.DynamicMethodDefinition.Trampoline<MonoMod.RuntimeDetour.ILHook+Context::Refresh>?560522894(object)
[ERROR @ 02:47:48 | IPA]   at HarmonyLib.Internal.RuntimeFixes.StackTraceFixes.OnILChainRefresh (System.Object self) [0x00000] in <bd235d216da54f80b5e45bcfb1bf9fda>:0 
[ERROR @ 02:47:48 | IPA]   at MonoMod.RuntimeDetour.ILHook.Apply () [0x00059] in <bbf7055c0fdb4018beccdadfe3e89f44>:0 
[ERROR @ 02:47:48 | IPA]   at HarmonyLib.Public.Patching.ManagedMethodPatcher.DetourTo (System.Reflection.MethodBase replacement) [0x00047] in <bd235d216da54f80b5e45bcfb1bf9fda>:0 
[ERROR @ 02:47:48 | IPA]    --- End of inner exception stack trace ---
[ERROR @ 02:47:48 | IPA]   at HarmonyLib.PatchClassProcessor.ReportException (System.Exception exception, System.Reflection.MethodBase original) [0x00045] in <bd235d216da54f80b5e45bcfb1bf9fda>:0 
[ERROR @ 02:47:48 | IPA]   at HarmonyLib.PatchClassProcessor.Patch () [0x00095] in <bd235d216da54f80b5e45bcfb1bf9fda>:0 
[ERROR @ 02:47:48 | IPA]   at HarmonyLib.Harmony.<PatchAll>b__11_0 (System.Type type) [0x00007] in <bd235d216da54f80b5e45bcfb1bf9fda>:0 
[ERROR @ 02:47:48 | IPA]   at HarmonyLib.CollectionExtensions.Do[T] (System.Collections.Generic.IEnumerable`1[T] sequence, System.Action`1[T] action) [0x00014] in <bd235d216da54f80b5e45bcfb1bf9fda>:0 
[ERROR @ 02:47:48 | IPA]   at HarmonyLib.Harmony.PatchAll (System.Reflection.Assembly assembly) [0x00006] in <bd235d216da54f80b5e45bcfb1bf9fda>:0 
[ERROR @ 02:47:48 | IPA]   at BetterSongList.Plugin.OnApplicationStart () [0x00011] in D:\VR\Beat Saber\source\BeatSaber_BetterSongList\Plugin.cs:27 
```

Reply:
Because there's no instance. State machines are structs.

Reply:
Just got the patch fixed. 🎉